File: src/green_moon_2d/gm_scene.nim
## This module is part of GreenMoon2D: https://github.com/willi-kappler/green_moon_2d
##
## Written by Willi Kappler, License: MIT
##
## This Nim library allows you to write 2D games, it uses Naylib (Raylib) internally.
##
## This module contains the code that handles all the scenes.
##

# Nim std imports
import std/json
import std/options
from std/strformat import fmt

# Local imports
import gm_log

type
    GMScene* = ref object of RootObj
        name: string

    GMSceneManager = object
        scenes: seq[GMScene]
        sceneStack: seq[string]
        currentScene: uint32

    GMSceneNotFoundError* = object of CatchableError
    GMSceneAlreadyExistsError* = object of CatchableError
    GMSceneIsActiveError* = object of CatchableError
    GMSceneStackEmptyError* = object of CatchableError

var GMGlobScenes: GMSceneManager

proc error_log2(filename: string, line: int, message: string, exceptn: typedesc) =
    gmLoggerError(fmt("{filename} - {line}: {message}"))
    raise newException(exceptn, message)

template error_log(message: string, exceptn: typedesc) =
    let pos = instantiationInfo()
    error_log2(pos.filename, pos.line, message, exceptn)

# GMScene:

method gmEnter*(self: var GMScene) {.base.} =
    ## This method can be implemented when the scene is entered.
    discard

method gmDraw*(self: var GMScene) {.base.} =
    ## This method must be implemented in order to draw the scene.
    quit("You must override this method: gm")

method gmDrawAfter*(self: var GMScene) {.base.} =
    ## This method can be implemented when things need to be drawn after all gfx objects.
    discard

method gmUpdate*(self: var GMScene) {.base.} =
    ## This method must be implemented in order to update the scene.
    quit("You must override this method: gm")

method gmUpdateAfter*(self: var GMScene) {.base.} =
    ## This method can be implemented when things need to be updated after all objects.
    discard

method gmCustom*(self: var GMScene, data: JsonNode): JsonNode {.base.} =
    ## This method can be implemented to send or receive user defined data.
    return newJNull()

# GMSceneManager:

proc gmFindSceneIndex(name: string): Option[uint32] =
    ## Return the index of a given scene (by name) or none if no such scene was found.
    for (i, s) in GMGlobScenes.scenes.pairs():
        if s.name == name:
            return some(uint32(i))

    return none(uint32)

proc gmDrawCurrentScene*() =
    ## Calls the gmDraw() method on the current active scene.
    GMGlobScenes.scenes[GMGlobScenes.currentScene].gmDraw()

proc gmDrawCurrentSceneAfter*() =
    ## Calls the gmDrawAfter() method on the current active scene.
    GMGlobScenes.scenes[GMGlobScenes.currentScene].gmDrawAfter()

proc gmDrawScene*(name: string) =
    ## Calls the gmDraw() method on the scene with the given name.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        GMGlobScenes.scenes[i].gmDraw()
    else:
        error_log(fmt("Can't draw scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmUpdateCurrentScene*() =
    ## Calls the gmUpdate() method on the current active scene.
    GMGlobScenes.scenes[GMGlobScenes.currentScene].gmUpdate()

proc gmUpdateCurrentSceneAfter*() =
    ## Calls the gmUpdateAfter() method on the current active scene.
    GMGlobScenes.scenes[GMGlobScenes.currentScene].gmUpdateAfter()

proc gmUpdateScene*(name: string) =
    ## Calls the gmUpdate() method on the scene with the given name.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        GMGlobScenes.scenes[i].gmUpdate()
    else:
        error_log(fmt("Can't update scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmEnterScene*(name: string) =
    ## Calls the gmEnter() method on the scene with the given name.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        GMGlobScenes.scenes[i].gmEnter()
    else:
        error_log(fmt("Can't enter scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmAddScene*(scene: GMScene, name: string) =
    ## Adds a new scene to the list of scenes.
    scene.name = name
    let idx = gmFindSceneIndex(scene.name)

    if idx.isSome():
        error_log(fmt("Can't add scene: {scene.name}, a scene with that name already exists!"), GMSceneAlreadyExistsError)
    else:
        GMGlobScenes.scenes.add(scene)

proc gmRemoveScene*(name: string) =
    ## Removes a scene with the given name.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        if i != GMGlobScenes.currentScene:
            # Remove scene
            GMGlobScenes.scenes.del(i)
        else:
            error_log(fmt("Can't remove current active scene: {name}, index: {i}!"), GMSceneIsActiveError)
    else:
        error_log(fmt("Can't remove scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmChangeScene*(name: string) =
    ## Changes the active scnene to the one with the given name.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        if i != GMGlobScenes.currentScene:
            GMGlobScenes.currentScene = i
            GMGlobScenes.scenes[i].gmEnter()
        else:
            error_log(fmt("Scene is already active: {name}, index: {i}!"), GMSceneIsActiveError)
    else:
        error_log(fmt("Can't change to scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmReplaceScene*(name: string, scene: GMScene) =
    ## Replace the scenen given by the name with the new given one.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        GMGlobScenes.scenes[i] = scene
    else:
        error_log(fmt("Can't replace scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmPushAndChangeScene*(name: string) =
    ## Pushes the current scene onto a stack and change to the given scene.
    ## This is useful if for example you want to show a menu during the game.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        if i != GMGlobScenes.currentScene:
            let currentSceneName = GMGlobScenes.scenes[GMGlobScenes.currentScene].name
            GMGlobScenes.sceneStack.add(currentSceneName)
            GMGlobScenes.currentScene = i
            GMGlobScenes.scenes[i].gmEnter()
        else:
            error_log(fmt("Can't push scene, it is already active: {name}, index: {i}!"), GMSceneIsActiveError)
    else:
        error_log(fmt("Can't push scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmPopAndChangeScene*() =
    ## Pops the last scene from the stack and changes to it.
    if GMGlobScenes.sceneStack.len() > 0:
        let previousSceneName = GMGlobScenes.sceneStack.pop()
        gmChangeScene(previousSceneName)
    else:
        error_log(fmt("Can't pop scene: scene stack is empty"), GMSceneStackEmptyError)

proc gmSceneCustomMessage*(name: string, data: JsonNode): JsonNode =
    ## Sends a message with custom data to the given scene.
    let idx = gmFindSceneIndex(name)

    if idx.isSome():
        let i = idx.get()
        return GMGlobScenes.scenes[i].gmCustom(data)
    else:
        error_log(fmt("Can't send cutom data to scene: {name}, it was not found!"), GMSceneNotFoundError)

proc gmInitSceneManager*() =
    GMGlobScenes.scenes = @[]
    GMGlobScenes.sceneStack = @[]
    GMGlobScenes.currentScene = 0

File: src/green_moon_2d/gm_scene.py
# This file is part of GreenMoon2D, a 2D game engine for Python
# written by Willi Kappler, MIT license.
#
# See: https://github.com/willi-kappler/green_moon_2d

"""
This module defines the GMSceneManager and the GMScene base class.
"""

from typing import Any

import logging
logger = logging.getLogger(__name__)


class GMScene:
    """
    This is the base class for a user defined scene.
    You have to provide a scene name and imlement the update() and draw() methods accordingly.
    """

    def __init__(self, name: str):
        """
        :param name: The name of the new scene. It must be unique.
        """

        logger.debug(f"Create a new GMScene with name: {name}.")

        self.name: str = name
        self.on_stack: bool = False
        self.initialized: bool = False
        self.custom_property: dict[str, Any] = {}

    def update(self) -> None:
        """
        This method is called once per frame and is used to update the
        internal state of the scene. You have to implement this method.
        """

        pass

    def draw(self) -> None:
        """
        This method is called once per frame and is used to draw the
        whole scene. You have to implement this method.
        """

        pass

    def enter(self) -> None:
        """
        This method is called every time this scene becomes the current
        scene. You don't have to implement this mwthod.
        """

        pass

    def leave(self) -> None:
        """
        This method is called every time this scene is no longer the
        current scene. You don't have to implement this mwthod.
        """

        pass

    def send_message(self, custom_message: Any):
        """
        This method can be used to send a custom message from one
        scene (usually the currently active one) to another scene.
        """

        pass


class GMSceneManager:
    """
    This class manages all the scenes that are used in the engine.
    """

    def __init__(self):
        logger.debug("Create a new GMSceneManager.")

        self.scenes: dict[str, GMScene] = {}
        self.current_scene: GMScene = GMScene("empty")
        self.scene_stack: list[GMScene] = []

    def update(self) -> None:
        """
        Updates the current scene. This method is called by the engine once per frame.
        """

        self.current_scene.update()

    def draw(self) -> None:
        """
        Draws the current scene. This method is called by the engine once per frame.

        :param context: The current game context.
        """

        self.current_scene.draw()

    def add_scene(self, scene: GMScene) -> None:
        """
        Adds a new scene to the scene manager. If a scene with the
        same name already exists it will be replaced.

        :param scene: The new scene to be added.
        """

        assert isinstance(scene, GMScene), "GMSceneManager.add_scene(), new scene must be a subclass of GMScene"

        logger.debug(f"Add a new scene with name: {scene.name}.")

        self.scenes[scene.name] = scene

    def delete_scene(self, name: str) -> None:
        """
        Deletes the scene with the given name.

        :param name: The name of the scene to be deleted.
        :raise KeyError: if the scene with the given name is not found.
        """

        logger.debug(f"Delete scene with name: {name}.")

        del self.scenes[name]

    def change_to_scene(self, name: str) -> None:
        """
        Change the current scene to the given one.

        :param name: The name of the scene to be deleted.
        :raise KeyError: if the scene with the given name is not found.
        """

        logger.debug(f"Change to scene with name: {name}.")

        self.current_scene.leave()
        self.current_scene = self.scenes[name]
        self.current_scene.enter()

    def push_and_change(self, name: str) -> None:
        """
        Push the current scene onto the scene stack and change to the given scene.

        :param name: The name of the scene to be deleted.
        :raise KeyError: if the scene with the given name is not found.
        """

        logger.debug("Push current scene and change to scene with name: {name}.")

        self.scene_stack.append(self.current_scene)
        self.current_scene.on_stack = True
        self.change_to_scene(name)

    def pop_and_change(self) -> None:
        """
        Pops a scene from the scene stack and change to it.

        :raise ValueError: if the stack is empty.
        """

        logger.debug("Pop scene from stack and change to it.")

        self.current_scene.leave()
        self.current_scene = self.scene_stack.pop()
        self.current_scene.on_stack = False
        self.current_scene.enter()

    def start_scene(self, name: str) -> None:
        """
        Sets the first scene to start the game with.

        :param name: The name of the scene to be deleted.
        :raise KeyError: if the scene with the given name is not found.
        """

        self.current_scene = self.scenes[name]
        self.current_scene.enter()

    def update_stack_top(self):
        """
        Update the scene that is on top of the stack.
        """

        self.scene_stack[-1].update()

    def draw_stack_top(self):
        """
        Draw the scene that is on top of the stack.
        """

        self.scene_stack[-1].draw()

    def update_scene(self, name: str):
        """
        Update the specific scene given by the name.

        :param name: The name of the scene to be updated.
        """

        self.scenes[name].update()

    def draw_scene(self, name: str):
        """
        Draw the specific scene given by the name.

        :param name: The name of the scene to be drawn.
        """

        self.scenes[name].draw()

    def send_message(self, name: str, custom_message: Any):
        """
        Sends a custom message to the scene given by the name.

        :param name: The name of the scene to send the custom message to.
        :param custom_message: The custom message that is sent to the scene.
        """

        self.scenes[name].send_message(custom_message)



